src/bio/bio_config.c

```diff
@@ -0,0 +1,532 @@
+	struct spdk_json_val *subsystems;
+	/* Current subsystem array position in "subsystems" array */
+	struct spdk_json_val *subsystems_it;
+
+	/* Current subsystem name */
+	struct spdk_json_val *subsystem_name;
+
+	/* Current "config" array */
+	struct spdk_json_val *config;
+	/* Current config position in "config" array */
+	struct spdk_json_val *config_it;
+
+	/* Whole configuration file read and parsed. */
+	size_t json_data_size;
+	char *json_data;
+
+	size_t values_cnt;
+	struct spdk_json_val *values;
+};
+
+static int
+cap_string(const struct spdk_json_val *val, void *out)
+{
+	const struct spdk_json_val **vptr = out;
+
+	if (val->type != SPDK_JSON_VAL_STRING) {
+		return -DER_INVAL;
+	}
+
+	*vptr = val;
+	return 0;
+}
+
+static int
+cap_object(const struct spdk_json_val *val, void *out)
+{
+	const struct spdk_json_val **vptr = out;
+
+	if (val->type != SPDK_JSON_VAL_OBJECT_BEGIN) {
+		return -DER_INVAL;
+	}
+
+	*vptr = val;
+	return 0;
+}
+
+
+static int
+cap_array_or_null(const struct spdk_json_val *val, void *out)
+{
+	const struct spdk_json_val **vptr = out;
+
+	if (val->type != SPDK_JSON_VAL_ARRAY_BEGIN && val->type != SPDK_JSON_VAL_NULL) {
+		return -DER_INVAL;
+	}
+
+	*vptr = val;
+	return 0;
+}
+
+static struct spdk_json_val *
+json_value(struct spdk_json_val *key)
+{
+	return key->type == SPDK_JSON_VAL_NAME ? key + 1 : NULL;
+}
+
+static struct spdk_json_object_decoder
+subsystem_decoders[] = {
+	{"subsystem", offsetof(struct json_config_ctx, subsystem_name), cap_string},
+	{"config", offsetof(struct json_config_ctx, config), cap_array_or_null}
+};
+
+struct
+config_entry {
+	char			*method;
+	struct spdk_json_val	*params;
+};
+
+static struct spdk_json_object_decoder
+jsonrpc_cmd_decoders[] = {
+	{"method", offsetof(struct config_entry, method), spdk_json_decode_string},
+	{"params", offsetof(struct config_entry, params), cap_object, true}
+};
+
+static int
+is_addr_in_allowlist(char *pci_addr, const struct spdk_pci_addr *allowlist,
+		     int num_allowlist_devices)
+{
+	int			i;
+	struct spdk_pci_addr    tmp;
+
+	if (spdk_pci_addr_parse(&tmp, pci_addr) != 0) {
+		D_ERROR("invalid transport address %s", pci_addr);
+		return -DER_INVAL;
+	}
+
+	for (i = 0; i < num_allowlist_devices; i++) {
+		if (spdk_pci_addr_compare(&tmp, &allowlist[i]) == 0) {
+			return 1;
+		}
+	}
+
+	return 0;
+}
+
+/*
+ * Convert a transport id in the BDF form of "5d0505:01:00.0" or something
+ * similar to the VMD address in the form of "0000:5d:05.5" that can be parsed
+ * by DPDK.
+ *
+ * \param dst String to be populated as output.
+ * \param src Input bdf.
+ */
+static int
+traddr_to_vmd(char *dst, const char *src)
+{
+	char traddr_tmp[SPDK_NVMF_TRADDR_MAX_LEN + 1];
+	char vmd_addr[SPDK_NVMF_TRADDR_MAX_LEN + 1] = "0000:";
+	char *ptr;
+	const char ch = ':';
+	char addr_split[3];
+	int position, iteration;
+	int n;
+
+	n = snprintf(traddr_tmp, SPDK_NVMF_TRADDR_MAX_LEN, "%s", src);
+	if (n < 0 || n > SPDK_NVMF_TRADDR_MAX_LEN) {
```

Comment:
The second case isn't actually possible, is it? Assuming snprintf is implemented to spec

Reply:
done